tests/frequentist_stats/test_linear_models.py

```diff
@@ -0,0 +1,510 @@
+
+
+def test_summarize_joint():
+    """Validates the structure of the comparative results object,
+    tests for accuracy of reported values are found above, in the
+    test__extract_<>_uplifts tests"""
+    _, alphas, _, model_df, column_label = _make_test_model()
+
+    actual = mafslm.summarize_joint(
+        model_df,
+        column_label,
+        ["control", "treatment-a", "treatment-b"],
+        alphas,
+        "treatment-a",
+    )
+
+    expected_index = [
+        ("abs_uplift", "0.5"),
+        ("abs_uplift", "exp"),
+        ("abs_uplift", "0.005"),
+        ("abs_uplift", "0.995"),
+        ("abs_uplift", "0.025"),
+        ("abs_uplift", "0.975"),
+        ("rel_uplift", "0.5"),
+        ("rel_uplift", "exp"),
+        ("rel_uplift", "0.005"),
+        ("rel_uplift", "0.995"),
+        ("rel_uplift", "0.025"),
+        ("rel_uplift", "0.975"),
+    ]
+
+    for branch in ["control", "treatment-b"]:
+        for key in expected_index:
+            assert key in actual[branch].index
+        assert len(actual[branch].index) == len(expected_index)
+
+
+def test__make_model_df():
+    # test removal of nulls
+    y = [1, 2, 3, 4, None, np.nan]
+    branch = ["control", "control", "treatment", "treatment", "control", "treatment"]
+    df_in = pd.DataFrame({"y": y, "branch": branch})
+    df_actual = mafslm._make_model_df(df_in, "y")
+    df_expected = pd.DataFrame({"branch": branch[:4], "y": [1.0, 2.0, 3.0, 4.0]})
+    pd.testing.assert_frame_equal(df_actual, df_expected)
+
+    # test removal of nulls with covariate
+    y = [1, 2, 3, 4, None, np.nan]
+    y2 = [1, 2, None, np.nan, 3, 4]
+    branch = ["control", "control", "treatment", "treatment", "control", "treatment"]
+    df_in = pd.DataFrame({"y": y, "branch": branch, "y2": y2})
+    df_actual = mafslm._make_model_df(df_in, "y", covariate_col_label="y2")
+    df_expected = pd.DataFrame(
+        {"branch": branch[:2], "y": [1.0, 2.0], "y2": [1.0, 2.0]}
+    )
+    pd.testing.assert_frame_equal(df_actual, df_expected)
+
+    # test thresholding
+    y = list(range(100))
+    branch = ["control"] * 100
+    df_in = pd.DataFrame({"y": y, "branch": branch})
+    df_actual = mafslm._make_model_df(df_in, "y", threshold_quantile=0.95)
+    df_expected = pd.DataFrame(
+        {
+            "branch": branch,
+            "y": [float(_y) for _y in y[:-5] + [np.quantile(y, 0.95)] * 5],
+        }
+    )
+    pd.testing.assert_frame_equal(df_actual, df_expected)
+
+    # test covariate & thresholding
+    y2 = list(range(100, 200))
+    branch = ["control"] * 100
+    df_in = pd.DataFrame({"y": y, "branch": branch, "y2": y2})
+    df_actual = mafslm._make_model_df(
+        df_in, "y", covariate_col_label="y2", threshold_quantile=0.95
+    )
+    df_expected = pd.DataFrame(
+        {
+            "branch": branch,
+            "y": [float(_y) for _y in y[:-5] + [np.quantile(y, 0.95)] * 5],
+            "y2": [float(_y) for _y in y2[:-5] + [np.quantile(y2, 0.95)] * 5],
```

Comment:
Why isn't this just list(np.quantile(y, .95)) ?

Reply:
Because the top 5 elements are clipped to the 95th percentile

Reply:
Another way to write this might be `np.clip(y, 0, np.quantile(y, 0.95)).astype(float)`

Reply:
AHHH okay I get it